### PR TITLE
Refactor redeemCode and support a Provider-wide EnrichSessionState method

### PR DESCRIPTION
**Pull request by NickMeves**
Split out session enrichment from OAuth Redeem functionality in `redeemCode`

Add EnrichSessionState provider method to unify use cases previous done with `GetUserName` & `GetEmailAddress`. Begin refactoring migrating few `GetUserName` providers to `EnrichSessionState` as a starting point.

## Motivation and Context

Codeclimate refactoring.

Simplify the provider interface and give an appropriate method to enrich sessions after Oauth2 redeem step. (A lot of providers seem to have hacked in AuthZ logic unrelated to emails into `GetEmailAddress` since it was the only viable hook).

## How Has This Been Tested?

New unit tests added.

GitHub provider test suite cleaned up to fix test bugs. 

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: providers/github_test.go
Comment (NickMeves) [inline providers/github_test.go:127]: I'm assuming the test wanted to test `NotVerified` like the name implies, hence the change in test behavior.

File: providers/github_test.go
Comment (NickMeves) [inline providers/github_test.go:224]: This test was broken before -- The error catch was catching an error in the `testGitHubBackend` by having an empty `[]string{}` for `/repo/oauth2-proxy/oauth2-proxy`.

I disagree with the underlying behavior of `hasRepo` -- `getEmail` raises an error and doesn't set an Email on an error in this method or a false return.  In the `false` case, the email that used to be returned was `""` and `nil` error was raised.

This refactor now at least makes it so `""` won't be returned and clobber a `session.Email` that might already be set. But I wonder if it is best to return an actual error? (I imagine this worked in the past to restrict access accidentally by setting the `session.Email` to "" which failed the email domain validation.

**Comment by NickMeves** [comment]
The `GetEmailAddress` method I renamed as `getEmail` is the source of the CodeClimate issues.  It is a giant monster and needs refactoring.  But it should wait until `Authorize` in #797 is merged -- alot of the bloat is from it doing all the authorization logic that didn't belong in `GetEmailAddress`.

I'll wait until a later PR to take on that refactoring.

File: providers/gitlab.go
Comment (JoelSpeed) [inline providers/gitlab.go:184]: Is this covered by the main email verification?

File: providers/provider_default.go
Comment (JoelSpeed) [inline providers/provider_default.go:105]: Does this PR remove all use of `GetEmailAddress`? Do we need to keep this in?

File: providers/github_test.go
Comment (JoelSpeed) [reply providers/github_test.go:224]: To clarify, if `hasRepo` fails, ie the user has failed the authorization check, what happens presently?

**Review by JoelSpeed — COMMENTED** [review]
Looks good, left a couple of comments for you when you're back from vacation!

File: providers/gitlab.go
Comment (NickMeves) [reply providers/gitlab.go:184]: Yup - this code seemed completely redundant to our email validator check that happens right after the `redeemCode` section that used to perform this check.

File: providers/provider_default.go
Comment (NickMeves) [reply providers/provider_default.go:105]: Still a bunch of `GetEmailAddress` uses unfortunately.  There were only 2 `GetUserName` uses, so I refactored those now while still keeping the PR small.

My goal is to go `GetEmailAddress` over time and have the few authorization related PRs for Keycloak and Gitlab coming in clean this up as a start.

File: providers/github_test.go
Comment (NickMeves) [reply providers/github_test.go:224]: https://github.com/oauth2-proxy/oauth2-proxy/blob/8be97f25e7276d2d46f002199f9eda345fc0800d/providers/github.go#L402

It looks like it will return `"", err` is either `hasRepo` errors or returns `false`.  This is fine at the moment, but not great -- and got accidental security from from our email validator not allowing empty emails.  But that will change with this PR: https://github.com/oauth2-proxy/oauth2-proxy/pull/770

Comment (NickMeves) [reply providers/github_test.go:224]: Even though it goes beyond  a refactor, I'm thinking we need to make `hasRepo` returning false return an error so the security logic is completely contained here appropriately.

Comment (NickMeves) [reply providers/github_test.go:224]: This pattern is all over the place in this provider:

```
err != nil || !ok {
  return err
}
```

is all over the place. A lot of `nil` errors are potentially returned in auth codepaths.  Punting for now, but this will need to be refactored later once we have the better `Authorization` scheme.

**Comment by NickMeves** [comment]
@JoelSpeed Bump on this one when you are available.

I want to get this and #797 soon since we have a few Provider refactor PRs from the community for Keycloak & Gitlab waiting on these 2 to merge.

File: CHANGELOG.md
Comment (JoelSpeed) [inline CHANGELOG.md:29]: We have duplicate entries here by the looks of it

**Review by JoelSpeed — COMMENTED** [review]
If we can get the changelog fixed I'm happy for this to merge

File: CHANGELOG.md
Comment (NickMeves) [reply CHANGELOG.md]: Good catch - I even had the wrong issue # as the link text 😅 

I rebased and fixed it

**Review by JoelSpeed — APPROVED** [review]
LGTM, thanks @NickMeves

**Pull request merged**
